test/DeployCampaign.integration.test.ts

```diff
@@ -0,0 +1,315 @@
+  let paidLongSubHash : string;
+
+  const mintAmount = ethers.utils.parseEther("10000000");
+
+  const domains = [shortDomain, mediumDomain, longDomain];
+
+  before(async () => {
+    [deployer, zeroVault, domainAddressMock, userA, userB, userC, userD, userE, userF] = await hre.ethers.getSigners();
+
+    // Reads `ENV_LEVEL` environment variable to determine rules to be enforced
+    config = getConfig(deployer, zeroVault);
+
+    config.mockMeowToken = hre.network.name === "hardhat";
+    const campaign = await runZnsCampaign({ config, dbVersion: "1.0.0" });
+
+    zns = campaign.state.contracts;
+
+    //  CurvePricer, stake, open
+    distConfig = {
+      pricerContract: zns.curvePricer.address,
+      paymentType: 1,
+      accessType: 1,
+    };
+
+    users = [
+      userA,
+      userB,
+      userC,
+      userD,
+      userE,
+      userF,
+    ];
+
+    freeShortSubHash = await zns.subRegistrar.hashWithParent(shortHash, freeShortSubdomain);
+    freeMediumSubHash = await zns.subRegistrar.hashWithParent(mediumHash, freeMediumSubdomain);
+    freeLongSubHash = await zns.subRegistrar.hashWithParent(longHash, freeLongSubdomain);
+
+    paidShortSubHash = await zns.subRegistrar.hashWithParent(shortHash, paidShortSubdomain);
+    paidMediumSubHash = await zns.subRegistrar.hashWithParent(mediumHash, paidMediumSubdomain);
+    paidLongSubHash = await zns.subRegistrar.hashWithParent(longHash, paidLongSubdomain);
+
+    await approveBulk(users, zns);
+
+    // Give the user funds
+    if (config.mockMeowToken) {
+      await mintBulk(
+        users,
+        mintAmount,
+        zns
+      );
+    }
+  });
+
+  it("Successfully mints TLDs with varying length", async () => {
+    // Confirm the domains are available
+    expect(await zns.registry.exists(shortHash)).to.be.false;
+    expect(await zns.registry.exists(mediumHash)).to.be.false;
+    expect(await zns.registry.exists(longHash)).to.be.false;
+
+    // Get domain prices, including stake and protocol fee
+    const [priceShort, priceMedium, priceLong] = await getPriceBulk(domains, zns);
+
+    logger.info(`Price of ${shortDomain} is ${priceShort.toString()}`);
+    logger.info(`Price of ${mediumDomain} is ${priceMedium.toString()}`);
+    logger.info(`Price of ${longDomain} is ${priceLong.toString()}`);
+
+    const balanceBeforePromises = [
+      zns.meowToken.balanceOf(userA.address),
+      zns.meowToken.balanceOf(userB.address),
+      zns.meowToken.balanceOf(userC.address),
+    ];
+
+    const [balanceBeforeA, balanceBeforeB, balanceBeforeC ]= await Promise.all(balanceBeforePromises);
+
+    // 1. Register root domains
+    // Note that this calls `setPriceConfig` internally for each TLD minted so we can also mint subdomains
+    await registerRootDomainBulk(
+      users,
+      domains,
+      domainAddressMock.address,
+      "https://zns.domains/", // tokenUri
+      distConfig,
+      config.rootPriceConfig,
+      zns
+    );
+
+    const balanceAfterPromises = [
```

Comment:
it can be the same array as before.

Reply:
I don't think this is true. I tried this to confirm and it only returns the before values when we don't have a second "after" array even if they are all the same promises.

When using a single promise array it fails the test with this error, and when using two different arrays (containing the same promises) it passes the tests. Must be another JavaScript behind the scenes thing, trying to save time and can see that the promises in the array are already resolved and so even with `await Promise.all()` running a second time, it just returns the resolved values
 
![image](https://github.com/zer0-os/zNS/assets/5976499/25941c2f-91f1-448e-a393-77f062e7b00c)